File: python/linkedlist.py

def main():
    values = [2,4,6,8,10]
    linkedlist = LinkedList(values)
    print(linkedlist.to_string(lambda x: str(x)))
    linkedlist.add(123)
    print(linkedlist.to_string(lambda x: str(x)))
    print('index 0', linkedlist.get(0))
    print('index 5', linkedlist.get(5))
    linkedlist.remove(5)
    linkedlist.remove(0)
    print(linkedlist.size(), linkedlist.to_string(lambda x: str(x)))


class LinkedList:
    def __init__(self, array=None):
        self.length = 0
        self.head = None

        if array != None:
            for value in array:
                self.add(value)

    
    def get(self, index):
        pos = 0
        node = self.head
        while node != None and pos < index:
            node = node.next
            pos+=1
        if node == None:
            raise IndexError("attempted to access index that does not exist")
        return node.value


    def add(self, value):
        new_node = LinkedListNode(value)
        if self.head == None:
            self.head = new_node
        else:
            last_node = self.head
            while last_node != None:
                if last_node.next == None:
                    break
                else:
                    last_node = last_node.next
            last_node.next = new_node
        self.length+=1


    def remove(self, index):
        prev_node = None
        node = self.head
        value_to_return = None

        for pos in range(1, index+1, 1):
            if node == None or (node.next == None and pos == index):
                raise IndexError('attempted to remove item from linked list at invalid index')
            prev_node = node
            node = node.next
        
        value_to_return = node.value

        if prev_node == None:
            self.head = node.next
        else:
            prev_node.next = node.next
        self.length-=1

        return value_to_return


    def size(self):
        return self.length


    def to_string(self, formatter):
        elems = []
        last_node = self.head
        while last_node != None:
            elems.append(formatter(last_node.value))
            last_node = last_node.next
        return '[' + ','.join(elems) + ']'
    

class LinkedListNode:
    def __init__(self, value):
        self.value = value
        self.next = None


if __name__ == '__main__':
    main()

File: javascript/src/linkedlist.ts
interface List<T> {
  get(index: number): T | null;
  add(value: T): void;
  remove(index: number): T | null;
  size(): number;
  toString(formatter: (value: T) => string): string;
}

export class LinkedList<T> implements List<T> {
  private length: number;
  private head: ListNode<T> | null;

  constructor(array?: T[]) {
    this.length = 0;
    this.head = null;

    !!array &&
      array.forEach((element) => {
        this.add(element);
      });
  }

  get(index: number): T | null {
    let pos = 0;
    let node = this.head;
    while (node !== null && pos < index) {
      node = node.next;
      pos++;
    }
    if (node === null) {
      throw new Error(
        `index-out-of-bounds accessing index ${index} from linked list`
      );
    }
    return node.value;
  }

  add(value: T) {
    const newNode = new ListNode<T>(value);
    if (this.head === null) {
      this.head = newNode;
    } else {
      let lastNode = this.head;
      while (lastNode != null) {
        if (lastNode.next === null) {
          break;
        } else {
          lastNode = lastNode.next;
        }
      }
      lastNode.next = newNode;
    }
    this.length++;
  }

  remove(index: number): T | null {
    const err = () => {
      throw new Error(
        `index-out-of-bounds: attempted to remove item from linked list at index ${index}`
      );
    };
    let prevNode: ListNode<T> | null = null;
    let node: ListNode<T> | null = this.head;
    let valueToReturn: T | null = null;

    for (let pos = 1; pos <= index; pos++) {
      if (node === null || (node.next === null && pos === index)) {
        return err();
      }
      prevNode = node;
      node = node.next;
    }

    if (node === null) {
      return err();
    }

    valueToReturn = node.value;

    if (prevNode === null) {
      this.head = node.next;
    } else {
      prevNode.next = node.next;
    }
    this.length--;

    return valueToReturn;
  }

  size(): number {
    return this.length;
  }

  toString(formatter: (value: T) => string): string {
    const elements: string[] = [];
    let node = this.head;
    while (node !== null) {
      elements.push(formatter(node.value));
      node = node.next;
    }
    return "[" + elements.join(",") + "]";
  }
}

class ListNode<T> {
  value: T;
  next: ListNode<T> | null;

  constructor(value: T, next?: ListNode<T>) {
    this.value = value;
    this.next = next ?? null;
  }
}
